File: src/tibo_radar/predictor.py
"""Explainable probability model for goodwill resets and banked resets."""

from __future__ import annotations

import math
import re
from datetime import datetime, timedelta, timezone
from typing import Any, Iterable
from zoneinfo import ZoneInfo

from .client import SourceBundle
from .models import Evidence, PredictionSnapshot, RadarEvent


BEIJING = ZoneInfo("Asia/Shanghai")
UTC = timezone.utc
DIRECT_FUTURE = re.compile(
    r"\b(will|tomorrow|later today|next hour|within (?:an? )?hour|lands? (?:at|around|by)|soon)\b",
    re.IGNORECASE,
)
RESET_WORDS = re.compile(r"\b(reset|reseting|resetting|banked reset)\b", re.IGNORECASE)
GLOBAL_WORDS = re.compile(
    r"\b(all paid|all users|everyone|every codex|every .* user|global)\b",
    re.IGNORECASE,
)
TARGETED_WORDS = re.compile(
    r"affected|eligible|some users|some plus|who used one|without access",
    re.IGNORECASE,
)
TEASER_WORDS = re.compile(
    r"hold on to your codex|burn (?:as much )?usage|you know what.s coming|who says it won.t reset|new milestone",
    re.IGNORECASE,
)
COMPLETED_WORDS = re.compile(
    r"all reset|have reset|has been reset|reset propagated|it is done|reset for everyone",
    re.IGNORECASE,
)


def parse_time(value: Any) -> datetime | None:
    if not isinstance(value, str) or not value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=UTC)
    return parsed.astimezone(UTC)


def clamp_probability(value: float) -> int:
    return max(1, min(99, round(value)))


def union_probability(first: int, second: int) -> int:
    result = 1 - (1 - first / 100) * (1 - second / 100)
    return clamp_probability(result * 100)


def _freshness_time(payloads: Iterable[dict[str, Any]]) -> datetime | None:
    candidates: list[datetime] = []
    for payload in payloads:
        for key in ("updated_at", "fetched_at", "checked_at", "newest_post_at"):
            parsed = parse_time(payload.get(key))
            if parsed:
                candidates.append(parsed)
    return max(candidates) if candidates else None


def _event_time(event: dict[str, Any]) -> datetime | None:
    for key in ("effective_at", "announced_at", "observed_at", "date"):
        parsed = parse_time(event.get(key))
        if parsed:
            return parsed
    return None


def _likely_window(forecast: dict[str, Any]) -> str:
    time_window = forecast.get("time_window") or {}
    start = time_window.get("start_hour")
    end = time_window.get("end_hour")
    timezone = time_window.get("timezone")
    if not isinstance(start, int) or not isinstance(end, int) or timezone != "UTC":
        return "北京时间凌晨至上午（历史高发段）"
    anchor = datetime(2026, 1, 1, start, tzinfo=UTC).astimezone(BEIJING)
    end_anchor = datetime(2026, 1, 1, end, tzinfo=UTC).astimezone(BEIJING)
    return f"北京时间 {anchor:%H}:00–{end_anchor:%H}:00"


def _latest_events(events: list[dict[str, Any]], limit: int = 5) -> list[RadarEvent]:
    ordered = sorted(events, key=lambda event: _event_time(event) or datetime.min.replace(tzinfo=UTC), reverse=True)
    output: list[RadarEvent] = []
    seen: set[str] = set()
    for event in ordered:
        event_id = str(event.get("id") or "")
        if not event_id or event_id in seen:
            continue
        seen.add(event_id)
        text = str(event.get("text") or event.get("summary") or "").strip()
        if not text:
            continue
        output.append(
            RadarEvent(
                event_id=event_id,
                kind=str(event.get("group") or event.get("type") or "signal"),
                title=text,
                occurred_at=(_event_time(event) or None).isoformat() if _event_time(event) else None,
                source_url=event.get("url") or event.get("source_url"),
                state=event.get("banked_state") or event.get("announcement_state"),
            )
        )
        if len(output) >= limit:
            break
    return output


class Predictor:
    """Blend a calibrated public baseline with small, visible event modifiers."""

    def predict(
        self, bundle: SourceBundle, now: datetime | None = None
    ) -> PredictionSnapshot:
        now = (now or datetime.now(UTC)).astimezone(UTC)
        forecast = bundle.payloads.get("forecast", {})
        timeline = bundle.payloads.get("timeline", {})
        status = bundle.payloads.get("status", {})
        openai_status = bundle.payloads.get("openai_status", {})
        events = list(timeline.get("events") or [])

        probabilities = forecast.get("probabilities") or {}
        global_24 = float(probabilities.get("rounded_24h") or 24)
        global_48 = float(probabilities.get("rounded_48h") or 43)
        banked_24 = 8.0
        banked_48 = 15.0
        affected_user_banked: int | None = None
        evidence: list[Evidence] = []

        cadence = forecast.get("cadence") or {}
        median_days = cadence.get("recent_median_days")
        age_days = forecast.get("age_days")
        if isinstance(median_days, (int, float)) and isinstance(age_days, (int, float)):
            evidence.append(
                Evidence(
                    "历史节奏",
                    f"距上次重置 {age_days:.1f} 天；近期典型间隔 {median_days:.1f} 天。",
                    "baseline",
                    0,
                    "https://codex-reset.com/timeline",
                )
            )

        last_reset = parse_time(forecast.get("last_reset_at"))
        if last_reset and now - last_reset < timedelta(hours=24):
            global_24 -= 8
            global_48 -= 4
            evidence.append(Evidence("重置冷却", "上次全局重置不足 24 小时。", "negative", -8))

        post_reset_events = [
            event for event in events
            if (_event_time(event) and (not last_reset or _event_time(event) > last_reset))
        ]
        post_reset_events.sort(key=lambda event: _event_time(event) or now, reverse=True)

        explicit_future: dict[str, Any] | None = None
        strong_teaser: dict[str, Any] | None = None
        latest_banked: dict[str, Any] | None = None
        for event in post_reset_events:
            text = str(event.get("text") or event.get("summary") or "")
            kind = str(event.get("group") or event.get("type") or "")
            if kind == "credits" and latest_banked is None:
                latest_banked = event
            if (
                kind == "reset"
                and RESET_WORDS.search(text)
                and DIRECT_FUTURE.search(text)
                and not COMPLETED_WORDS.search(text)
                and explicit_future is None
            ):
                explicit_future = event
            elif TEASER_WORDS.search(text) and strong_teaser is None:
                strong_teaser = event

        if explicit_future:
            text = str(explicit_future.get("text") or explicit_future.get("summary") or "")
            floor = 88 if GLOBAL_WORDS.search(text) else 72
            global_24 = max(global_24, floor)
            global_48 = max(global_48, min(96, floor + 7))
            evidence.append(
                Evidence(
                    "Tibo 明确预告",
                    text[:180],
                    "positive",
                    floor,
                    explicit_future.get("url"),
                )
            )
        elif strong_teaser:
            global_24 += 18
            global_48 += 12
            evidence.append(
                Evidence(
                    "Tibo 强暗示",
                    str(strong_teaser.get("text") or strong_teaser.get("summary") or "")[:180],
                    "positive",
                    18,
                    strong_teaser.get("url"),
                )
            )

        if latest_banked:
            text = str(latest_banked.get("text") or latest_banked.get("summary") or "")
            occurred_at = _event_time(latest_banked)
            recent = occurred_at is not None and now - occurred_at <= timedelta(hours=48)
            targeted = bool(TARGETED_WORDS.search(text))
            broad = bool(GLOBAL_WORDS.search(text)) and not targeted
            state = str(latest_banked.get("banked_state") or "")
            if recent and targeted:
                affected_user_banked = 94
                evidence.append(
                    Evidence(
                        "定向补发",
                        "Tibo 已承诺补发给故障时段内受影响的用户；这不会提高普发概率。",
                        "targeted",
                        0,
                        latest_banked.get("url"),
                    )
                )
            elif recent and broad and state in {"announced", "arriving"}:
                banked_24 = max(banked_24, 94)
                banked_48 = max(banked_48, 98)
                evidence.append(
                    Evidence(
                        "重置卡已预告",
                        text[:180],
                        "positive",
                        86,
                        latest_banked.get("url"),
                    )
                )
            elif recent and broad:
                banked_24 = max(banked_24, 72)
                banked_48 = max(banked_48, 84)

        incident = self._recent_incident(openai_status, status, now)
        if incident:
            active = str(incident.get("status")) not in {"resolved", "postmortem", "completed"}
            delta = 12 if active else 5
            global_24 += delta
            global_48 += max(3, delta - 4)
            evidence.append(
                Evidence(
                    "Codex 状态事件",
                    str(incident.get("name") or "近期 Codex 服务异常"),
                    "positive" if active else "context",
                    delta,
                    incident.get("shortlink") or incident.get("source_url"),
                )
            )

        global_24_i = clamp_probability(global_24)
        global_48_i = max(global_24_i, clamp_probability(global_48))
        banked_24_i = clamp_probability(banked_24)
        banked_48_i = max(banked_24_i, clamp_probability(banked_48))
        combined_24 = union_probability(global_24_i, banked_24_i)
        combined_48 = max(combined_24, union_probability(global_48_i, banked_48_i))

        if combined_24 >= 85:
            level, label = "red", "很可能"
        elif combined_24 >= 65:
            level, label = "orange", "高关注"
        elif combined_24 >= 45:
            level, label = "yellow", "开始关注"
        else:
            level, label = "green", "正常使用"

        data_updated = _freshness_time(bundle.payloads.values())
        stale = bool(bundle.cache_fallbacks)
        if data_updated and now - data_updated > timedelta(hours=4):
            stale = True

        confidence = str(forecast.get("confidence") or "low")
        if explicit_future or banked_24_i >= 90:
            confidence = "high"
        elif stale:
            confidence = "low"
        confidence_note = str(
            forecast.get("confidence_note")
            or "历史样本较少，概率只用于安排用量。"
        )
        if stale:
            confidence_note = "部分数据来自缓存；请把当前概率视为低置信度。"

        return PredictionSnapshot(
            generated_at=now.isoformat(),
            global_24h=global_24_i,
            global_48h=global_48_i,
            banked_24h=banked_24_i,
            banked_48h=banked_48_i,
            combined_24h=combined_24,
            combined_48h=combined_48,
            affected_user_banked_24h=affected_user_banked,
            confidence=confidence,
            confidence_note=confidence_note,
            level=level,
            level_label=label,
            likely_window=_likely_window(forecast),
            last_reset_at=last_reset.isoformat() if last_reset else None,
            data_updated_at=data_updated.isoformat() if data_updated else None,
            stale=stale,
            evidence=evidence[:6],
            latest_events=_latest_events(events),
            source_errors=bundle.errors,
        )

    @staticmethod
    def _recent_incident(
        openai_status: dict[str, Any], fallback_status: dict[str, Any], now: datetime
    ) -> dict[str, Any] | None:
        incidents = list(openai_status.get("incidents") or [])
        incidents.extend(fallback_status.get("incidents") or [])
        candidates: list[tuple[datetime, dict[str, Any]]] = []
        seen: set[str] = set()
        for incident in incidents:
            incident_id = str(incident.get("id") or "")
            if incident_id in seen:
                continue
            seen.add(incident_id)
            text = f"{incident.get('name', '')} {incident.get('body', '')}".lower()
            codex_related = incident.get("codex_related") is True or any(
                term in text for term in ("codex", "work mode", "usage limit")
            )
            if not codex_related:
                continue
            at = parse_time(
                incident.get("resolved_at")
                or incident.get("updated_at")
                or incident.get("started_at")
            )
            if at and now - at <= timedelta(hours=36):
                candidates.append((at, incident))
        if not candidates:
            return None
        candidates.sort(key=lambda item: item[0], reverse=True)
        return candidates[0][1]

File: native/Sources/TiboRadarApp/Predictor.swift
import Foundation

struct Predictor: Sendable {
  private let directFutureTerms = [
    " will ", "tomorrow", "later today", "next hour", "within an hour",
    "within a hour", "lands at", "lands around", "lands by", "soon",
  ]
  private let resetTerms = ["reset", "resetting", "banked reset"]
  private let globalTerms = [
    "all paid", "all users", "everyone", "every codex", "global",
  ]
  private let targetedTerms = [
    "affected", "eligible", "some users", "some plus", "who used one",
    "without access", "affected time window",
  ]
  private let teaserTerms = [
    "hold on to your codex", "burn as much usage", "burn usage",
    "you know what's coming", "who says it won't reset", "new milestone",
  ]
  private let completedTerms = [
    "all reset", "have reset", "has been reset", "reset propagated",
    "it is done", "reset for everyone",
  ]

  func predict(bundle: SourceBundle, now: Date = Date()) -> PredictionSnapshot {
    let forecast = bundle.payloads[.forecast]?.objectValue ?? [:]
    let timeline = bundle.payloads[.timeline]?.objectValue ?? [:]
    let feed = bundle.payloads[.feed]?.objectValue ?? [:]
    let status = bundle.payloads[.status]?.objectValue ?? [:]
    let openAIStatus = bundle.payloads[.openAIStatus]?.objectValue ?? [:]
    let events = mergedEvents(timeline: timeline, feed: feed)

    let probabilities = forecast.object("probabilities") ?? [:]
    var global24 = probabilities.number("rounded_24h") ?? 24
    var global48 = probabilities.number("rounded_48h") ?? 43
    var banked24 = 8.0
    var banked48 = 15.0
    var affectedUserBanked: Int?
    var evidence: [Evidence] = []

    let cadence = forecast.object("cadence") ?? [:]
    if let medianDays = cadence.number("recent_median_days"),
      let ageDays = forecast.number("age_days")
    {
      evidence.append(
        Evidence(
          id: "cadence",
          label: "历史节奏",
          detail: String(
            format: "距上次重置 %.1f 天；近期典型间隔 %.1f 天。",
            ageDays,
            medianDays
          ),
          category: "baseline",
          sourceURL: URL(string: "https://codex-reset.com/timeline")
        ))
    }

    let lastReset = parseDate(forecast.string("last_reset_at"))
    if let lastReset, now.timeIntervalSince(lastReset) < 24 * 60 * 60 {
      global24 -= 8
      global48 -= 4
      evidence.append(
        Evidence(
          id: "cooldown",
          label: "重置冷却",
          detail: "上次全局重置不足 24 小时。",
          category: "negative",
          delta: -8
        ))
    }

    let postResetEvents =
      events
      .filter { event in
        guard let date = eventDate(event) else { return false }
        return lastReset.map { date > $0 } ?? true
      }
      .sorted { (eventDate($0) ?? .distantPast) > (eventDate($1) ?? .distantPast) }

    var explicitFuture: [String: JSONValue]?
    var strongTeaser: [String: JSONValue]?
    var latestBanked: [String: JSONValue]?

    for event in postResetEvents {
      let text = eventText(event)
      let kind = eventKind(event)
      if ["credits", "banked"].contains(kind), latestBanked == nil {
        latestBanked = event
      }
      if kind == "reset",
        containsAny(text, resetTerms),
        containsAny(" \(text) ", directFutureTerms),
        !containsAny(text, completedTerms),
        explicitFuture == nil
      {
        explicitFuture = event
      } else if containsAny(text, teaserTerms), strongTeaser == nil {
        strongTeaser = event
      }
    }

    if let event = explicitFuture {
      let text = eventText(event)
      let floor = containsAny(text, globalTerms) ? 88 : 72
      global24 = max(global24, Double(floor))
      global48 = max(global48, Double(min(96, floor + 7)))
      evidence.append(
        Evidence(
          id: "explicit-future",
          label: "Tibo 明确预告",
          detail: clipped(text),
          category: "positive",
          delta: floor,
          sourceURL: eventURL(event)
        ))
    } else if let event = strongTeaser {
      global24 += 18
      global48 += 12
      evidence.append(
        Evidence(
          id: "strong-teaser",
          label: "Tibo 强暗示",
          detail: clipped(eventText(event)),
          category: "positive",
          delta: 18,
          sourceURL: eventURL(event)
        ))
    }

    if let event = latestBanked {
      let text = eventText(event)
      let recent = eventDate(event).map { now.timeIntervalSince($0) <= 48 * 60 * 60 } ?? false
      let targeted = containsAny(text, targetedTerms)
      let broad = containsAny(text, globalTerms) && !targeted
      let state = event.string("banked_state") ?? ""

      if recent, targeted {
        affectedUserBanked = 94
        evidence.append(
          Evidence(
            id: "targeted-compensation",
            label: "定向补发",
            detail: "Tibo 已承诺补发给故障时段内受影响的用户；这不会提高普发概率。",
            category: "targeted",
            sourceURL: eventURL(event)
          ))
      } else if recent, broad, ["announced", "arriving"].contains(state) {
        banked24 = max(banked24, 94)
        banked48 = max(banked48, 98)
        evidence.append(
          Evidence(
            id: "banked-announced",
            label: "重置卡已预告",
            detail: clipped(text),
            category: "positive",
            delta: 86,
            sourceURL: eventURL(event)
          ))
      } else if recent, broad {
        banked24 = max(banked24, 72)
        banked48 = max(banked48, 84)
      }
    }

    if let incident = recentIncident(
      openAIStatus: openAIStatus,
      fallbackStatus: status,
      now: now
    ) {
      let incidentStatus = incident.string("status") ?? ""
      let active = !["resolved", "postmortem", "completed"].contains(incidentStatus)
      let delta = active ? 12 : 5
      global24 += Double(delta)
      global48 += Double(max(3, delta - 4))
      evidence.append(
        Evidence(
          id: "recent-incident",
          label: "Codex 状态事件",
          detail: incident.string("name") ?? "近期 Codex 服务异常",
          category: active ? "positive" : "context",
          delta: delta,
          sourceURL: URL(
            string: incident.string("shortlink") ?? incident.string("source_url") ?? "")
        ))
    }

    let global24Value = clamp(global24)
    let global48Value = max(global24Value, clamp(global48))
    let banked24Value = clamp(banked24)
    let banked48Value = max(banked24Value, clamp(banked48))
    let combined24 = union(global24Value, banked24Value)
    let combined48 = max(combined24, union(global48Value, banked48Value))

    let level: ProbabilityLevel =
      if combined24 >= 85 {
        .red
      } else if combined24 >= 65 {
        .orange
      } else if combined24 >= 45 {
        .yellow
      } else {
        .green
      }

    let dataUpdatedAt = freshnessDate(in: bundle.payloads.values)
    let isStale =
      !bundle.cacheFallbacks.isEmpty
      || dataUpdatedAt.map { now.timeIntervalSince($0) > 4 * 60 * 60 } == true

    var confidence = forecast.string("confidence") ?? "low"
    if explicitFuture != nil || banked24Value >= 90 {
      confidence = "high"
    } else if isStale {
      confidence = "low"
    }
    let sourceNote =
      forecast.string("confidence_note")
      ?? "历史样本较少，概率只用于安排用量。"

    return PredictionSnapshot(
      generatedAt: now,
      global24h: global24Value,
      global48h: global48Value,
      banked24h: banked24Value,
      banked48h: banked48Value,
      combined24h: combined24,
      combined48h: combined48,
      affectedUserBanked24h: affectedUserBanked,
      confidence: confidence,
      confidenceNote: isStale ? "部分数据来自缓存；请把当前概率视为低置信度。" : sourceNote,
      level: level,
      likelyWindow: likelyWindow(forecast),
      lastResetAt: lastReset,
      dataUpdatedAt: dataUpdatedAt,
      isStale: isStale,
      evidence: Array(evidence.prefix(6)),
      latestEvents: latestEvents(events),
      sourceErrors: bundle.errors
    )
  }

  private func mergedEvents(
    timeline: [String: JSONValue],
    feed: [String: JSONValue]
  ) -> [[String: JSONValue]] {
    let raw = (timeline.array("events") ?? []) + (feed.array("events") ?? [])
    var seen = Set<String>()
    var merged: [[String: JSONValue]] = []
    for value in raw {
      guard let event = value.objectValue else { continue }
      let id = event.string("id") ?? ""
      guard !id.isEmpty, seen.insert(id).inserted else { continue }
      merged.append(event)
    }
    return merged
  }

  private func latestEvents(_ events: [[String: JSONValue]]) -> [RadarEvent] {
    events
      .sorted { (eventDate($0) ?? .distantPast) > (eventDate($1) ?? .distantPast) }
      .compactMap { event -> RadarEvent? in
        guard let id = event.string("id"), !id.isEmpty else { return nil }
        let title = eventText(event).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !title.isEmpty else { return nil }
        return RadarEvent(
          id: id,
          kind: eventKind(event),
          title: title,
          occurredAt: eventDate(event),
          sourceURL: eventURL(event),
          state: event.string("banked_state") ?? event.string("announcement_state")
        )
      }
      .prefix(5)
      .map { $0 }
  }

  private func recentIncident(
    openAIStatus: [String: JSONValue],
    fallbackStatus: [String: JSONValue],
    now: Date
  ) -> [String: JSONValue]? {
    let values =
      (openAIStatus.array("incidents") ?? [])
      + (fallbackStatus.array("incidents") ?? [])
    var seen = Set<String>()
    var candidates: [(Date, [String: JSONValue])] = []

    for value in values {
      guard let incident = value.objectValue else { continue }
      let id = incident.string("id") ?? ""
      guard seen.insert(id).inserted else { continue }
      let searchable = "\(incident.string("name") ?? "") \(incident.string("body") ?? "")"
        .lowercased()
      let related =
        incident.bool("codex_related") == true
        || ["codex", "work mode", "usage limit"].contains { searchable.contains($0) }
      guard related else { continue }
      let date = parseDate(
        incident.string("resolved_at")
          ?? incident.string("updated_at")
          ?? incident.string("started_at")
      )
      if let date, now.timeIntervalSince(date) <= 36 * 60 * 60 {
        candidates.append((date, incident))
      }
    }
    return candidates.max { $0.0 < $1.0 }?.1
  }

  private func eventDate(_ event: [String: JSONValue]) -> Date? {
    for key in ["effective_at", "announced_at", "observed_at", "at", "date"] {
      if let date = parseDate(event.string(key)) { return date }
    }
    return nil
  }

  private func eventText(_ event: [String: JSONValue]) -> String {
    event.string("text") ?? event.string("summary") ?? ""
  }

  private func eventKind(_ event: [String: JSONValue]) -> String {
    event.string("group") ?? event.string("type") ?? event.string("kind") ?? "signal"
  }

  private func eventURL(_ event: [String: JSONValue]) -> URL? {
    URL(string: event.string("url") ?? event.string("source_url") ?? "")
  }

  private func parseDate(_ value: String?) -> Date? {
    guard let value, !value.isEmpty else { return nil }
    let fractional = ISO8601DateFormatter()
    fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let date = fractional.date(from: value) { return date }
    let regular = ISO8601DateFormatter()
    regular.formatOptions = [.withInternetDateTime]
    if let date = regular.date(from: value) { return date }
    if value.count == 10 {
      let formatter = DateFormatter()
      formatter.calendar = Calendar(identifier: .gregorian)
      formatter.locale = Locale(identifier: "en_US_POSIX")
      formatter.timeZone = TimeZone(secondsFromGMT: 0)
      formatter.dateFormat = "yyyy-MM-dd"
      return formatter.date(from: value)
    }
    return nil
  }

  private func freshnessDate(in payloads: Dictionary<RadarSource, JSONValue>.Values) -> Date? {
    payloads.compactMap { payload in
      guard let object = payload.objectValue else { return nil }
      return ["updated_at", "fetched_at", "checked_at", "newest_post_at"]
        .compactMap { parseDate(object.string($0)) }
        .max()
    }.max()
  }

  private func likelyWindow(_ forecast: [String: JSONValue]) -> String {
    guard let window = forecast.object("time_window"),
      let start = window.int("start_hour"),
      let end = window.int("end_hour"),
      window.string("timezone") == "UTC"
    else {
      return "北京时间凌晨至上午"
    }
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = TimeZone(secondsFromGMT: 0)!
    let startDate = calendar.date(from: DateComponents(year: 2026, month: 1, day: 1, hour: start))!
    let endDate = calendar.date(from: DateComponents(year: 2026, month: 1, day: 1, hour: end))!
    let formatter = DateFormatter()
    formatter.locale = Locale(identifier: "zh_CN")
    formatter.timeZone = TimeZone(identifier: "Asia/Shanghai")
    formatter.dateFormat = "HH:mm"
    return "北京时间 \(formatter.string(from: startDate))–\(formatter.string(from: endDate))"
  }

  private func containsAny(_ text: String, _ terms: [String]) -> Bool {
    let lowered = text.lowercased()
    return terms.contains { lowered.contains($0) }
  }

  private func clipped(_ text: String, limit: Int = 180) -> String {
    String(text.prefix(limit))
  }

  private func clamp(_ value: Double) -> Int {
    max(1, min(99, Int(value.rounded())))
  }

  private func union(_ first: Int, _ second: Int) -> Int {
    let value = 1 - (1 - Double(first) / 100) * (1 - Double(second) / 100)
    return clamp(value * 100)
  }
}
